src/CodeGen_Vulkan_Dev.cpp

```diff
@@ -0,0 +1,2928 @@
+}
+
+template<typename T>
+SpvId CodeGen_Vulkan_Dev::SPIRV_Emitter::declare_constant_int(Type value_type, int64_t value) {
+    const T typed_value = (T)(value);
+    SpvId constant_id = builder.declare_constant(value_type, &typed_value);
+    builder.update_id(constant_id);
+    return constant_id;
+}
+
+template<typename T>
+SpvId CodeGen_Vulkan_Dev::SPIRV_Emitter::declare_constant_uint(Type value_type, uint64_t value) {
+    const T typed_value = (T)(value);
+    SpvId constant_id = builder.declare_constant(value_type, &typed_value);
+    builder.update_id(constant_id);
+    return constant_id;
+}
+
+template<typename T>
+SpvId CodeGen_Vulkan_Dev::SPIRV_Emitter::declare_constant_float(Type value_type, float value) {
+    const T typed_value = (T)(value);
+    SpvId constant_id = builder.declare_constant(value_type, &typed_value);
+    builder.update_id(constant_id);
+    return constant_id;
+}
+
+void CodeGen_Vulkan_Dev::SPIRV_Emitter::visit(const IntImm *imm) {
+    if (imm->type.bits() == 8) {
+        declare_constant_int<int8_t>(imm->type, imm->value);
+    } else if (imm->type.bits() == 16) {
+        declare_constant_int<int16_t>(imm->type, imm->value);
+    } else if (imm->type.bits() == 32) {
+        declare_constant_int<int32_t>(imm->type, imm->value);
+    } else if (imm->type.bits() == 64) {
+        declare_constant_int<int64_t>(imm->type, imm->value);
+    } else {
+        internal_error << "Vulkan backend currently only supports 8-bit, 16-bit, 32-bit or 64-bit signed integers!\n";
+    }
+}
+
+void CodeGen_Vulkan_Dev::SPIRV_Emitter::visit(const UIntImm *imm) {
+    if (imm->type.bits() == 8) {
+        declare_constant_uint<uint8_t>(imm->type, imm->value);
+    } else if (imm->type.bits() == 16) {
+        declare_constant_uint<uint16_t>(imm->type, imm->value);
+    } else if (imm->type.bits() == 32) {
+        declare_constant_uint<uint32_t>(imm->type, imm->value);
+    } else if (imm->type.bits() == 64) {
+        declare_constant_uint<uint64_t>(imm->type, imm->value);
+    } else {
+        internal_error << "Vulkan backend currently only supports 8-bit, 16-bit, 32-bit or 64-bit unsigned integers!\n";
+    }
+}
+
+void CodeGen_Vulkan_Dev::SPIRV_Emitter::visit(const StringImm *imm) {
+    SpvId constant_id = builder.declare_string_constant(imm->value);
+    builder.update_id(constant_id);
+}
+
+void CodeGen_Vulkan_Dev::SPIRV_Emitter::visit(const FloatImm *imm) {
+    if (imm->type.bits() == 16) {
+        if (imm->type.is_bfloat()) {
+            declare_constant_float<bfloat16_t>(imm->type, imm->value);
+        } else {
+            declare_constant_float<float16_t>(imm->type, imm->value);
+        }
+    } else if (imm->type.bits() == 32) {
+        declare_constant_float<float>(imm->type, imm->value);
+    } else if (imm->type.bits() == 64) {
+        declare_constant_float<double>(imm->type, imm->value);
+    } else {
+        internal_error << "Vulkan backend currently only supports 32-bit or 64-bit floats\n";
```

Comment:
This method seems to also handle bfloat16 and float16, is the error message incorrect?

Reply:
Good catch ... yes comment is out of date.  I'll fix this.